File: todo.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using Spectre.Console;

namespace TodoApp
{
    class Task
    {
        public int Id { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }
        public string Priority { get; set; } // high, medium, low
        public string Due { get; set; }
        public bool Completed { get; set; }
    }

    class Program
    {
        static List<Task> tasks = new List<Task>();
        static int nextId = 1;
        const string DataFile = "tasks.json";

        static void Main()
        {
            Load();
            while (true)
            {
                AnsiConsole.Clear();
                var choice = AnsiConsole.Prompt(
                    new SelectionPrompt<string>()
                        .Title("[bold yellow]=== Todo Manager ===[/]")
                        .PageSize(8)
                        .AddChoices(new[] {
                            "➕ Добавить задачу",
                            "📋 Список всех",
                            "✅ Выполнить / восстановить",
                            "🗑 Удалить",
                            "🔍 Фильтр (активные/выполненные)",
                            "💾 Выход"
                        }));
                switch (choice)
                {
                    case "➕ Добавить задачу": AddTask(); break;
                    case "📋 Список всех": ListAll(); break;
                    case "✅ Выполнить / восстановить": ToggleComplete(); break;
                    case "🗑 Удалить": DeleteTask(); break;
                    case "🔍 Фильтр (активные/выполненные)": Filter(); break;
                    case "💾 Выход": return;
                }
                AnsiConsole.MarkupLine("\n[grey]Нажмите любую клавишу...[/]");
                Console.ReadKey();
            }
        }

        static void Load()
        {
            if (File.Exists(DataFile))
            {
                string json = File.ReadAllText(DataFile);
                tasks = JsonSerializer.Deserialize<List<Task>>(json);
                if (tasks.Count > 0) nextId = tasks.Max(t => t.Id) + 1;
            }
            else tasks = new List<Task>();
        }

        static void Save() => File.WriteAllText(DataFile, JsonSerializer.Serialize(tasks));

        static void AddTask()
        {
            var title = AnsiConsole.Ask<string>("[green]Название:[/]");
            if (string.IsNullOrWhiteSpace(title)) { AnsiConsole.MarkupLine("[red]Ошибка[/]"); return; }
            var desc = AnsiConsole.Ask<string>("[grey]Описание:[/]");
            var priority = AnsiConsole.Prompt(
                new SelectionPrompt<string>().Title("Приоритет:").AddChoices("high", "medium", "low"));
            var due = AnsiConsole.Ask<string>("[grey]Дедлайн (ГГГГ-ММ-ДД, Enter пропустить):[/]");
            tasks.Add(new Task { Id = nextId++, Title = title, Description = desc, Priority = priority, Due = due, Completed = false });
            Save();
            AnsiConsole.MarkupLine("[green]✓ Добавлено[/]");
        }

        static void ListAll()
        {
            if (!tasks.Any()) { AnsiConsole.MarkupLine("[yellow]Нет задач[/]"); return; }
            var table = new Table();
            table.AddColumn("ID");
            table.AddColumn("Статус");
            table.AddColumn("Название");
            table.AddColumn("Приоритет");
            table.AddColumn("Дедлайн");
            foreach (var t in tasks)
            {
                string status = t.Completed ? "[green]✓[/]" : "[grey] [/]";
                string prioColor = t.Priority == "high" ? "red" : (t.Priority == "medium" ? "yellow" : "green");
                table.AddRow(t.Id.ToString(), status, t.Title, $"[{prioColor}]{t.Priority}[/]", t.Due ?? "-");
            }
            AnsiConsole.Render(table);
        }

        static void ToggleComplete()
        {
            int id = AnsiConsole.Ask<int>("ID задачи: ");
            var task = tasks.FirstOrDefault(t => t.Id == id);
            if (task != null) { task.Completed = !task.Completed; Save(); AnsiConsole.MarkupLine("[green]Статус изменён[/]"); }
            else AnsiConsole.MarkupLine("[red]Не найдено[/]");
        }

        static void DeleteTask()
        {
            int id = AnsiConsole.Ask<int>("ID для удаления: ");
            if (tasks.RemoveAll(t => t.Id == id) > 0) { Save(); AnsiConsole.MarkupLine("[green]Удалено[/]"); }
            else AnsiConsole.MarkupLine("[red]Не найдено[/]");
        }

        static void Filter()
        {
            var filter = AnsiConsole.Prompt(
                new SelectionPrompt<string>().Title("Фильтр:").AddChoices("Активные", "Выполненные"));
            var filtered = filter == "Активные" ? tasks.Where(t => !t.Completed) : tasks.Where(t => t.Completed);
            foreach (var t in filtered) AnsiConsole.MarkupLine($"{t.Id}: {t.Title}");
        }
    }
}

File: todo.rs
// dependencies in Cargo.toml:
// [dependencies]
// serde = { version = "1.0", features = ["derive"] }
// serde_json = "1.0"
// dialoguer = "0.11"
// chrono = "0.4"

use serde::{Serialize, Deserialize};
use std::fs;
use std::io;
use dialoguer::{Input, Select, Confirm};
use chrono::NaiveDate;

#[derive(Serialize, Deserialize, Clone)]
struct Task {
    id: u32,
    title: String,
    description: String,
    priority: String,
    due: Option<String>,
    completed: bool,
}

const DATA_FILE: &str = "tasks.json";

fn load_tasks() -> Vec<Task> {
    if let Ok(data) = fs::read_to_string(DATA_FILE) {
        serde_json::from_str(&data).unwrap_or_else(|_| Vec::new())
    } else {
        Vec::new()
    }
}

fn save_tasks(tasks: &Vec<Task>) {
    let data = serde_json::to_string_pretty(tasks).unwrap();
    fs::write(DATA_FILE, data).unwrap();
}

fn add_task(tasks: &mut Vec<Task>) {
    let title: String = Input::new().with_prompt("Название").interact_text().unwrap();
    if title.is_empty() { println!("Ошибка"); return; }
    let description: String = Input::new().with_prompt("Описание").default("".into()).interact_text().unwrap();
    let priority = ["Высокий", "Средний", "Низкий"];
    let prio_idx = Select::new().items(&priority).default(1).interact().unwrap();
    let priority_str = match prio_idx { 0 => "high", 1 => "medium", _ => "low" }.to_string();
    let due_str: String = Input::new().with_prompt("Дедлайн (ГГГГ-ММ-ДД, Enter пропустить)").allow_empty(true).interact_text().unwrap();
    let due = if due_str.is_empty() { None } else { Some(due_str) };
    let id = tasks.iter().map(|t| t.id).max().unwrap_or(0) + 1;
    tasks.push(Task { id, title, description, priority: priority_str, due, completed: false });
    save_tasks(tasks);
    println!("✅ Задача {} добавлена", id);
}

fn list_tasks(tasks: &Vec<Task>, filter: Option<&str>) {
    let filtered: Vec<&Task> = tasks.iter()
        .filter(|t| match filter {
            Some("active") => !t.completed,
            Some("completed") => t.completed,
            _ => true
        }).collect();
    if filtered.is_empty() { println!("Нет задач."); return; }
    for t in filtered {
        let status = if t.completed { "✓" } else { " " };
        let prio_sym = match t.priority.as_str() {
            "high" => "🔴", "medium" => "🟡", _ => "🟢"
        };
        let due = t.due.as_deref().unwrap_or("-");
        println!("[{}] {}: {} {} | {} | до {}", status, t.id, t.title, prio_sym, t.description, due);
    }
}

fn toggle_complete(tasks: &mut Vec<Task>) {
    let id: u32 = Input::new().with_prompt("ID задачи").interact_text().unwrap();
    if let Some(t) = tasks.iter_mut().find(|t| t.id == id) {
        t.completed = !t.completed;
        save_tasks(tasks);
        println!("Статус изменён");
    } else { println!("Не найдено"); }
}

fn delete_task(tasks: &mut Vec<Task>) {
    let id: u32 = Input::new().with_prompt("ID для удаления").interact_text().unwrap();
    let len = tasks.len();
    tasks.retain(|t| t.id != id);
    if tasks.len() < len { save_tasks(tasks); println!("Удалено"); }
    else { println!("Не найдено"); }
}

fn main() {
    let mut tasks = load_tasks();
    loop {
        println!("\n=== Rust Todo ===");
        let options = &["Добавить", "Список всех", "Активные", "Выполненные", "Выполнить/вернуть", "Удалить", "Выход"];
        let choice = Select::new().items(options).default(0).interact().unwrap();
        match choice {
            0 => add_task(&mut tasks),
            1 => list_tasks(&tasks, None),
            2 => list_tasks(&tasks, Some("active")),
            3 => list_tasks(&tasks, Some("completed")),
            4 => toggle_complete(&mut tasks),
            5 => delete_task(&mut tasks),
            _ => break,
        }
    }
}
